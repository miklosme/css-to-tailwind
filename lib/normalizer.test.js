const { parseCss } = require('./parsers');
const { createRounder } = require('./utils');
const normalizer = require('./normalizer');
const { setOptions } = require('./options');

const { getVariantFromSelector } = normalizer;

test('normalizer should handle custom spacing', async () => {
    setOptions({
        TAILWIND_CONFIG: {
            theme: {
                spacing: {
                    '36': '36px',
                    '72': '72px',
                    '144': '144px',
                },
            },
        },
    });

    const parsedCss = await parseCss(`
        div {
            padding: 4rem;
        }
    `);
    const normalizedCss = normalizer(parsedCss);

    expect(normalizedCss.base.div['padding-bottom']).not.toBe('64px'); // default config value
    expect(normalizedCss.base.div['padding-bottom']).toBe('72px');
    expect(normalizedCss).toMatchInlineSnapshot(`
        Object {
          "base": Object {
            "div": Object {
              "padding-bottom": "72px",
              "padding-left": "72px",
              "padding-right": "72px",
              "padding-top": "72px",
            },
          },
        }
    `);
});

const responsiveBreakpointNames = {
    640: 'sm',
    768: 'md',
    1024: 'lg',
    1280: 'xl',
};

const breakpointRounder = createRounder({
    breakpoints: Object.keys(responsiveBreakpointNames).sort((a, z) => a - z),
});

// TODO test variantsKey
test.each([
    ['div', null, { baseSelector: 'div', variants: [] }],
    ['div:hover', null, { baseSelector: 'div', variants: ['hover'] }],
    ['.foo:focus', null, { baseSelector: '.foo', variants: ['focus'] }],
    ['.bar::placeholder', null, { baseSelector: '.bar', variants: ['placeholder'] }],
    ['.baz:focus::placeholder', null, { baseSelector: '.baz', variants: ['focus', 'placeholder'] }],
    ['.foo::-ms-input-placeholder', null, { baseSelector: '.foo', variants: ['placeholder'] }],
    ['.foo::-moz-placeholder', null, { baseSelector: '.foo', variants: ['placeholder'] }],
    [
        'div p >.foo .bar:hover #id .baz:focus::placeholder',
        null,
        { baseSelector: 'div p > .foo .bar:hover #id .baz', variants: ['focus', 'placeholder'] },
    ],
    ['   p    div:hover    ', null, { baseSelector: 'p div', variants: ['hover'] }],
    ['.foo', '(min-width: 640px)', { baseSelector: '.foo', variants: ['sm'] }],
    ['.foo', '(min-width: 768px)', { baseSelector: '.foo', variants: ['md'] }],
    ['.foo', '(min-width: 1024px)', { baseSelector: '.foo', variants: ['lg'] }],
    ['.foo', '(min-width: 1280px)', { baseSelector: '.foo', variants: ['xl'] }],
    ['.foo:hover', '(min-width: 1280px)', { baseSelector: '.foo', variants: ['xl', 'hover'] }],
])("getVariantFromSelector('%s', '%s')", (selector, mediaRuleValue, expected) => {
    const { baseSelector, variants } = getVariantFromSelector({
        selector,
        mediaRuleValue,
        responsiveBreakpointNames,
        breakpointRounder,
    });
    expect(baseSelector).toEqual(expected.baseSelector);
    expect(variants).toEqual(expected.variants);
});

test('getVariantFromSelector should round breakpoints', () => {
    const { variants } = getVariantFromSelector({
        selector: '.foo',
        mediaRuleValue: '(min-width: 700px)',
        responsiveBreakpointNames,
        breakpointRounder,
    });
    expect(variants).toEqual(['sm']);
});

test('getVariantFromSelector should handle complex atrule media params', () => {
    const { variants } = getVariantFromSelector({
        selector: '.foo',
        mediaRuleValue: 'all and (min-width: 1111px) and print',
        responsiveBreakpointNames,
        breakpointRounder,
    });
    expect(variants).toEqual(['lg']);
});
