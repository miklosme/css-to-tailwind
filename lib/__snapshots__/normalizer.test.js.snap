// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizer should handle variants 2`] = `
Object {
  "base": Object {
    ".bar": Object {
      "padding-bottom": "24px",
      "padding-left": "80px",
      "padding-right": "80px",
      "padding-top": "24px",
      "position": "relative",
    },
    ".baz": Object {
      "background-image": "url(logo.svg)",
      "background-repeat": "no-repeat",
      "padding-bottom": "24px",
      "padding-left": "80px",
      "padding-right": "80px",
      "padding-top": "24px",
    },
    ".foo": Object {
      "border-bottom-width": "1px",
      "border-color": "rgba(250, 208, 208, 1)",
      "border-left-width": "1px",
      "border-radius": "4px",
      "border-right-width": "1px",
      "border-style": "solid",
      "border-top-width": "1px",
      "border-width": "1px",
      "color": "rgba(255, 255, 255, 1)",
      "margin-bottom": "24px",
      "padding-bottom": "24px",
      "padding-left": "80px",
      "padding-right": "80px",
      "padding-top": "24px",
      "position": "relative",
      "width": "100%",
    },
  },
  "focus": Object {
    ".foo": Object {
      "border-color": "rgba(226, 242, 240, 1)",
    },
  },
  "hover": Object {
    ".foo": Object {
      "background-color": "rgba(204, 204, 204, 1)",
    },
  },
  "placeholder": Object {
    ".bar": Object {
      "color": "rgba(230, 230, 230, 1)",
    },
  },
  "xl": Object {
    ".bar": Object {
      "margin-bottom": "40px",
      "padding-bottom": "48px",
      "padding-left": "96px",
      "padding-right": "96px",
      "padding-top": "48px",
    },
  },
}
`;
